iothub/device/src/Transport/AmqpClientConnectionPool.cs

```diff
@@ -0,0 +1,155 @@
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityIoTHubSas)
+            {
+                return GetIoTHubSasClientConnection(deviceClientEndpointIdentity);
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+        }
+        internal void RemoveClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+
+            if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySasSingle)
+            {
+                if (singleSasPool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    singleSasPool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from singleSasPool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityX509)
+            {
+                if (x509Pool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    x509Pool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from X509Pool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            if (Logging.IsEnabled) Logging.Exit(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+        }
+
+
+        private AmqpClientConnection GetIoTHubSasSingleClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (singleSasPool.ContainsKey(deviceClientEndpointIdentity))
+            {
+                // Error handling!!!
+                return null;
+            }
+            else
+            {
+                AmqpClientConnectionFactory amqpClientConnectionFactory = new AmqpClientConnectionFactory();
+
+                AmqpClientConnection amqpClientConnection = amqpClientConnectionFactory.Create(deviceClientEndpointIdentity, RemoveClientConnection);
+                if (singleSasPool.TryAdd(deviceClientEndpointIdentity, amqpClientConnection))
+                {
+                    return amqpClientConnection;
+                }
+                else
+                {
+                    // Error handling!!!
+                    return null;
```

Comment:
Should we close the `amqpClientConnection`?

Reply:
This is not relevant anymore., as the singleSasPool has been removed completely as I am implementing the Pooling. 

---
In reply to: [251626875](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251626875) [](ancestors = 251626875)